[Bug]: QueryCoord ut data race

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:
- Deployment mode(standalone or cluster):
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```


### Current Behavior

```
2022-05-12T07:22:31.9356598Z ==================
2022-05-12T07:22:31.9357075Z WARNING: DATA RACE
2022-05-12T07:22:31.9357635Z Write at 0x00c0027684a8 by goroutine 643:
2022-05-12T07:22:31.9358211Z   container/list.(*List).insert()
2022-05-12T07:22:31.9358800Z       /usr/local/go/src/container/list/list.go:98 +0x724
2022-05-12T07:22:31.9359395Z   container/list.(*List).insertValue()
2022-05-12T07:22:31.9359992Z       /usr/local/go/src/container/list/list.go:104 +0x4f9
2022-05-12T07:22:31.9360593Z   container/list.(*List).PushBack()
2022-05-12T07:22:31.9361185Z       /usr/local/go/src/container/list/list.go:155 +0x4d7
2022-05-12T07:22:31.9362199Z   github.com/milvus-io/milvus/internal/querycoord.(*channelUnsubscribeHandler).addUnsubscribeChannelInfo()
2022-05-12T07:22:31.9363109Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/channel_unsubscribe.go:117 +0x47b
2022-05-12T07:22:31.9363929Z   github.com/milvus-io/milvus/internal/querycoord.(*queryNodeCluster).setNodeState()
2022-05-12T07:22:31.9364749Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/cluster.go:557 +0x15a
2022-05-12T07:22:31.9365547Z   github.com/milvus-io/milvus/internal/querycoord.(*queryNodeCluster).stopNode()
2022-05-12T07:22:31.9366500Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/cluster.go:634 +0x1c7
2022-05-12T07:22:31.9367294Z   github.com/milvus-io/milvus/internal/querycoord.TestGetShardLeaders()
2022-05-12T07:22:31.9368110Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/impl_test.go:1689 +0xd6b
2022-05-12T07:22:31.9368626Z   testing.tRunner()
2022-05-12T07:22:31.9369196Z       /usr/local/go/src/testing/testing.go:1193 +0x202
2022-05-12T07:22:31.9369569Z 
2022-05-12T07:22:31.9383978Z Previous read at 0x00c0027684a8 by goroutine 516:
2022-05-12T07:22:31.9385049Z   container/list.(*List).Front()
2022-05-12T07:22:31.9385665Z       /usr/local/go/src/container/list/list.go:70 +0x60e
2022-05-12T07:22:31.9386698Z   github.com/milvus-io/milvus/internal/querycoord.(*channelUnsubscribeHandler).handleChannelUnsubscribeLoop()
2022-05-12T07:22:31.9387611Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/channel_unsubscribe.go:132 +0x5d7
2022-05-12T07:22:31.9387946Z 
2022-05-12T07:22:31.9391267Z Goroutine 643 (running) created at:
2022-05-12T07:22:31.9393523Z   testing.(*T).Run()
2022-05-12T07:22:31.9394227Z       /usr/local/go/src/testing/testing.go:1238 +0x5d7
2022-05-12T07:22:31.9394806Z   testing.runTests.func1()
2022-05-12T07:22:31.9395386Z       /usr/local/go/src/testing/testing.go:1511 +0xa6
2022-05-12T07:22:31.9395943Z   testing.tRunner()
2022-05-12T07:22:31.9396510Z       /usr/local/go/src/testing/testing.go:1193 +0x202
2022-05-12T07:22:31.9397627Z   testing.runTests()
2022-05-12T07:22:31.9398123Z       /usr/local/go/src/testing/testing.go:1509 +0x612
2022-05-12T07:22:31.9398694Z   testing.(*M).Run()
2022-05-12T07:22:31.9399252Z       /usr/local/go/src/testing/testing.go:1417 +0x3b3
2022-05-12T07:22:31.9400130Z   github.com/milvus-io/milvus/internal/querycoord.TestMain()
2022-05-12T07:22:31.9401050Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/query_coord_test.go:65 +0x49
2022-05-12T07:22:31.9401582Z   main.main()
2022-05-12T07:22:31.9402100Z       _testmain.go:823 +0x391
2022-05-12T07:22:31.9402453Z 
2022-05-12T07:22:31.9410533Z Goroutine 516 (running) created at:
2022-05-12T07:22:31.9411687Z   github.com/milvus-io/milvus/internal/querycoord.(*channelUnsubscribeHandler).start()
2022-05-12T07:22:31.9412575Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/channel_unsubscribe.go:152 +0x86
2022-05-12T07:22:31.9413364Z   github.com/milvus-io/milvus/internal/querycoord.(*QueryCoord).Start()
2022-05-12T07:22:31.9414163Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/query_coord.go:234 +0x116
2022-05-12T07:22:31.9415048Z   github.com/milvus-io/milvus/internal/querycoord.startQueryCoord()
2022-05-12T07:22:31.9416903Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/query_coord_test.go:108 +0x553
2022-05-12T07:22:31.9417709Z   github.com/milvus-io/milvus/internal/querycoord.TestGetShardLeaders()
2022-05-12T07:22:31.9418516Z       /go/src/github.com/milvus-io/milvus/internal/querycoord/impl_test.go:1639 +0x7e
2022-05-12T07:22:31.9419274Z   testing.tRunner()
2022-05-12T07:22:31.9419847Z       /usr/local/go/src/testing/testing.go:1193 +0x202
2022-05-12T07:22:31.9420408Z ==================
```

--- FAIL: TestGetShardLeaders (0.84s)

https://github.com/milvus-io/milvus/runs/6401120251?check_suite_focus=true

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Milvus Log

_No response_

### Anything else?

_No response_

--- a/internal/querycoord/channel_unsubscribe.go
+++ b/internal/querycoord/channel_unsubscribe.go
@@ -45,6 +45,7 @@ type channelUnsubscribeHandler struct {
 	kvClient *etcdkv.EtcdKV
 	factory  msgstream.Factory
 
+	mut          sync.RWMutex // mutex for channelInfos, since container/list is not goroutine-safe
 	channelInfos *list.List
 	downNodeChan chan int64
 
@@ -73,6 +74,13 @@ func newChannelUnsubscribeHandler(ctx context.Context, kv *etcdkv.EtcdKV, factor
 	return handler, nil
 }
 
+// appendUnsubInfo pushes unsub info safely
+func (csh *channelUnsubscribeHandler) appendUnsubInfo(info *querypb.UnsubscribeChannelInfo) {
+	csh.mut.Lock()
+	defer csh.mut.Unlock()
+	csh.channelInfos.PushBack(info)
+}
+
 // reloadFromKV reload unsolved channels to unsubscribe
 func (csh *channelUnsubscribeHandler) reloadFromKV() error {
 	log.Info("start reload unsubscribe channelInfo from kv")
@@ -86,7 +94,7 @@ func (csh *channelUnsubscribeHandler) reloadFromKV() error {
 		if err != nil {
 			return err
 		}
-		csh.channelInfos.PushBack(channelInfo)
+		csh.appendUnsubInfo(channelInfo)
 		csh.downNodeChan <- channelInfo.NodeID
 	}
 
@@ -102,11 +110,14 @@ func (csh *channelUnsubscribeHandler) addUnsubscribeChannelInfo(info *querypb.Un
 	}
 	// when queryCoord is restarted multiple times, the nodeID of added channelInfo may be the same
 	hasEnqueue := false
+	// reduce the lock range to iteration here, since `addUnsubscribeChannelInfo` is called one by one
+	csh.mut.RLock()
 	for e := csh.channelInfos.Back(); e != nil; e = e.Prev() {
 		if e.Value.(*querypb.UnsubscribeChannelInfo).NodeID == nodeID {
 			hasEnqueue = true
 		}
 	}
+	csh.mut.RUnlock()
 
 	if !hasEnqueue {
 		channelInfoKey := fmt.Sprintf("%s/%d", unsubscribeChannelInfoPrefix, nodeID)
@@ -114,7 +125,7 @@ func (csh *channelUnsubscribeHandler) addUnsubscribeChannelInfo(info *querypb.Un
 		if err != nil {
 			panic(err)
 		}
-		csh.channelInfos.PushBack(info)
+		csh.appendUnsubInfo(info)
 		csh.downNodeChan <- info.NodeID
 		log.Info("add unsubscribeChannelInfo to handler", zap.Int64("nodeID", info.NodeID))
 	}
@@ -129,7 +140,10 @@ func (csh *channelUnsubscribeHandler) handleChannelUnsubscribeLoop() {
 			log.Info("channelUnsubscribeHandler ctx done, handleChannelUnsubscribeLoop end")
 			return
 		case <-csh.downNodeChan:
+			csh.mut.RLock()
+			e := csh.channelInfos.Front()
 			channelInfo := csh.channelInfos.Front().Value.(*querypb.UnsubscribeChannelInfo)
+			csh.mut.RUnlock()
 			nodeID := channelInfo.NodeID
 			for _, collectionChannels := range channelInfo.CollectionChannels {
 				collectionID := collectionChannels.CollectionID
@@ -142,6 +156,10 @@ func (csh *channelUnsubscribeHandler) handleChannelUnsubscribeLoop() {
 				log.Error("remove unsubscribe channelInfo from etcd failed", zap.Int64("nodeID", nodeID))
 				panic(err)
 			}
+
+			csh.mut.Lock()
+			csh.channelInfos.Remove(e)
+			csh.mut.Unlock()
 			log.Info("unsubscribe channels success", zap.Int64("nodeID", nodeID))
 		}
 	}

